File: fsharp/robot-name/RobotName.fs
module RobotName

type Robot = { name: string }

let makeName() =
    let letters = ['A'..'Z']
    let random = new System.Random()
    sprintf "%c%c%03i" letters.[random.Next(26)] letters.[random.Next(26)] (random.Next(999))

let mkRobot() = { name = makeName() }

let name robot = robot.name

let reset robot = { robot with name = makeName() }

File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    private static readonly Random random = new Random();
    private static readonly List<string> used = new List<string>();
    private static readonly string alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private static string GetName ()
    {
        while(true)
        {
            var candidate = $"{alpha[random.Next(25)]}{alpha[random.Next(25)]}{random.Next(9)}{random.Next(9)}{random.Next(9)}";
            if(!used.Contains(candidate))
            {
                used.Add(candidate);
                return candidate;
            }
        }
    }

    public string Name { get; private set; } = GetName();

    public void Reset() => Name = GetName();
}